File: package/ios/HybridUserComModule.swift
import Foundation
import UIKit
import NitroModules
import UserComSDK

class HybridUserComModule: HybridUserComModuleSpec {
    
    func initialize(config: UserComModuleConfig) throws -> NitroModules.Promise<Void> {
        NSLog("[UserCom] HybridUserCom native initializing")
        
        let domain = config.domain.trimmingCharacters(in: .whitespacesAndNewlines)
        let host = domain.replacingOccurrences(of: "^https?://", with: "", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let sdk = UserSDK(
            application: UIApplication.shared,
            apiKey: config.apiKey,
            baseURL: host,
            shouldTrackActivities: config.trackAllActivities ?? false
        )
        
        let promise = Promise<Void>()
        sdk.ping { success, error in
            if let error = error {
                promise.reject(withError: error)
            } else if !success {
                promise.reject(withError: NSError(domain: "User.com ping failed", code: 1))
            } else {
                promise.resolve()
            }
        }
        return promise
    }
    
    func registerUser(userData: UserComModuleUserData) throws -> NitroModules.Promise<UserComModuleRegisterUserResponse> {
        let promise = Promise<UserComModuleRegisterUserResponse>()
        
        guard let sdk: UserSDK = UserSDK.default else {
            promise.reject(withError: NSError(domain: "SDK is not initialized, call initialize() first", code: 0))
            return promise
        }
        
        var standardData: [UserSDK.UserDataKey: String?] = [.userId: userData.id]
        if let firstName = userData.firstName { standardData[.firstName] = firstName }
        if let lastName = userData.lastName { standardData[.lastName] = lastName }
        if let email = userData.email { standardData[.email] = email }
        if let phoneNumber = userData.phoneNumber { standardData[.phone] = phoneNumber }

        sdk.setUserData(standardData) { success, error in
            if let error = error {
                promise.reject(withError: error)
                return
            }
            guard success else {
                promise.reject(withError: NSError(domain: "User.com identify failed", code: 1))
                return
            }

            guard let attributes = userData.attributes, !attributes.isEmpty else {
                promise.resolve(withResult: UserComModuleRegisterUserResponse.first(NullType.null))
                return
            }

            sdk.setCustomUserData(attributes.mapValues { $0 }) { attributesSuccess, attributesError in
                if let attributesError = attributesError {
                    promise.reject(withError: attributesError)
                } else if !attributesSuccess {
                    promise.reject(withError: NSError(domain: "User.com attributes update failed", code: 1))
                } else {
                    promise.resolve(withResult: UserComModuleRegisterUserResponse.first(NullType.null))
                }
            }
        }
        
        return promise
    }
    
    func logout() throws -> NitroModules.Promise<Void> {
        let promise = Promise<Void>()
        
        guard let sdk: UserSDK = UserSDK.default else {
            promise.reject(withError: NSError(domain: "SDK is not initialized, call initialize() first", code: 0))
            return promise
        }
        
        sdk.logout(fcmToken: nil) { success, error in
            if let error = error {
                promise.reject(withError: error)
            } else if !success {
                promise.reject(withError: NSError(domain: "User.com logout failed", code: 1))
            } else {
                promise.resolve()
            }
        }
        
        return promise
    }
    
    private func mapToProductEventType(_ eventType: UserComProductEventType) -> UserSDK.EventType {
        switch eventType {
            case .addtocart: return .addToCart
            case .purchase: return .purchase
            case .liking: return .liking
            case .addtoobservation: return .addToObservation
            case .order: return .order
            case .reservation: return .reservation
            case .return: return .return
            case .view: return .view
            case .click: return .click
            case .detail: return .detail
            case .add: return .add
            case .remove: return .remove
            case .checkout: return .checkout
            case .checkoutoption: return .checkoutOption
            case .refund: return .refund
            case .promoclick: return .promoClick
        }
    }

    func sendProductEvent(productId: String, eventType: UserComProductEventType, params: NitroModules.AnyMap?) throws -> NitroModules.Promise<Void> {

        guard let sdk: UserSDK = UserSDK.default else {
            return Promise.rejected(
                withError: NSError(domain: "SDK is not initialized, call initialize() first", code: 0)
            )
        }

        let promise = Promise<Void>()
        
        let sdkEventType = mapToProductEventType(eventType)
        sdk.sendProductEvent(productId, eventType: sdkEventType, params: params?.toDictionary().compactMapValues{ $0 }) { success, error in
            if let error = error {
                promise.reject(withError: error)
            } else if !success {
                promise.reject(withError: NSError(domain: "User.com product event failed", code: 1))
            } else {
                promise.resolve()
            }
        }

        return promise
    }
    
    func sendCustomEvent(eventName: String, data: NitroModules.AnyMap) throws -> NitroModules.Promise<Void> {
        
        guard let sdk: UserSDK = UserSDK.default else {
            return Promise.rejected(
                withError: NSError(domain: "SDK is not initialized, call initialize() first", code: 0)
            )
        }
        
        let promise = Promise<Void>()
        sdk.sendEvent(with: eventName, params: data.toDictionary().compactMapValues{ $0 }) { success, error in
            if let error = error {
                promise.reject(withError: error)
            } else if !success {
                promise.reject(withError: NSError(domain: "User.com custom event failed", code: 1))
            } else {
                promise.resolve()
            }
        }
        
        return promise
    }
    
    func sendScreenEvent(screenName: String) throws -> NitroModules.Promise<Void> {
        guard let sdk: UserSDK = UserSDK.default else {
            return Promise.rejected(
                withError: NSError(domain: "SDK is not initialized, call initialize() first", code: 0)
            )
        }
        
        let promise = Promise<Void>()
        sdk.trackScreen(with: screenName) { success, error in
            if let error = error {
                promise.reject(withError: error)
                return
            }
            if success {
                promise.resolve()
            } else {
                promise.reject(withError: NSError(domain: "User.com screen event failed", code: 1))
            }
        }
        
        return promise
    }
}

File: package/android/src/main/java/com/margelo/nitro/usercom/HybridUserComModule.kt
package com.margelo.nitro.usercom

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.AnyMap
import com.margelo.nitro.core.Promise
import com.margelo.nitro.core.resolve
import com.margelo.nitro.core.resolved
import com.user.sdk.UserCom
import com.user.sdk.customer.Customer
import com.user.sdk.customer.CustomerUpdateCallback
import com.user.sdk.customer.RegisterResponse
import com.user.sdk.events.ProductEventType

@Keep
@DoNotStrip
class HybridUserComModule : HybridUserComModuleSpec() {

    private val defaultInitTimeout = 10000L

    private fun buildCustomer(customerData: UserComModuleUserData): Customer {
        val customer = Customer()
        customer.id(customerData.id)
        customerData.email?.let { customer.email(it) }
        customerData.firstName?.let { customer.firstName(it) }
        customerData.lastName?.let { customer.lastName(it) }
        customerData.phoneNumber?.let { customer.attr("phone_number", it) }
        customerData.attributes?.forEach { (key, value) ->
            val attrVal = value.asFirstOrNull()
                ?: value.asSecondOrNull() ?: value.asThirdOrNull()
            when (attrVal) {
                is String -> customer.attr(key, attrVal)
                is Boolean -> customer.attr(key, attrVal)
                is Double -> {
                    require(attrVal.isFinite() && attrVal % 1.0 == 0.0 && attrVal >= Int.MIN_VALUE && attrVal <= Int.MAX_VALUE) {
                        "Android User.com SDK does not support decimal contact attributes: $key"
                    }
                    customer.attr(key, attrVal.toInt())
                }
            }
        }
        return customer
    }

    override fun initialize(config: UserComModuleConfig): Promise<Unit> {
        val application = NitroModules.applicationContext?.currentActivity?.application
            ?: return Promise.rejected(
                Throwable("Application context is not available")
            )

        val instance = try {
            UserCom.getInstance()
        } catch (_: Throwable) {
            null
        }
        val promise = Promise<Unit>()
        if (instance != null) {
            Log.d("HybridUserComModule", "UserCom SDK already initialized")
            promise.resolve()
            return promise
        }

        val handler = Handler(Looper.getMainLooper())

        val timeoutRunnable = Runnable {
            promise.reject(Throwable("UserCom SDK have not been initialized within given timeout"))
        }

        handler.postDelayed(timeoutRunnable, config.initTimeoutMs?.toLong() ?: defaultInitTimeout)

        val initHandler = object : UserCom.OnSdkInitializedListener {
            override fun onSdkInitialized(p0: Customer) {
                handler.removeCallbacks(timeoutRunnable)
                promise.resolve()
            }

            override fun onUserRegistrationFailed() {
                handler.removeCallbacks(timeoutRunnable)
                promise.reject(Throwable("User registration failed. Check native exceptions."))
            }
        }

        // Android SDK expects a URL; iOS SDK expects a host. Accept either form from JS.
        val domain = config.domain.trim().trimEnd('/')
        val baseUrl = if (domain.startsWith("https://") || domain.startsWith("http://")) {
            "$domain/"
        } else {
            "https://$domain/"
        }
        try {
            val builder =
                UserCom.Builder(application, config.apiKey, config.integrationsApiKey, baseUrl)
            builder.setOnSdkInitializedListener(initHandler)
            config.trackAllActivities?.let { builder.trackAllActivities(it) }
            config.openLinksInChromeCustomTabs?.let { builder.openLinksInChromeCustomTabs(it) }
            config.defaultCustomer?.let { builder.setDefaultCustomer(buildCustomer(it)) }
            builder.build()
        } catch (error: Throwable) {
            handler.removeCallbacks(timeoutRunnable)
            promise.reject(error)
        }

        return promise
    }

    override fun registerUser(userData: UserComModuleUserData): Promise<UserComModuleRegisterUserResponse> {
        val promise = Promise<UserComModuleRegisterUserResponse>()

        val instance = try {
            UserCom.getInstance()
        } catch (_: Throwable) {
            promise.reject(Throwable("SDK is not initialized, call initialize() first"))
            return promise
        }

        val customer = try {
            buildCustomer(userData)
        } catch (error: Throwable) {
            return Promise.rejected(error)
        }

        instance.register(customer, object : CustomerUpdateCallback {
                override fun onSuccess(p0: RegisterResponse) {
                    promise.resolve(UserComModuleRegisterUserResponse.create(p0.key))
                }

                override fun onFailure(p0: Throwable) {
                    promise.reject(p0)
                }
            })
        return promise
    }

    override fun logout(): Promise<Unit> {
        val instance = try {
            UserCom.getInstance()
        } catch (_: Throwable) {
            return Promise.rejected(Throwable("SDK is not initialized, call initialize() first"))
        }
        instance.logout()

        return Promise.resolved()
    }

    override fun sendProductEvent(
        productId: String,
        eventType: UserComProductEventType,
        params: AnyMap?
    ): Promise<Unit> {
        val instance = try {
            UserCom.getInstance()
        } catch (_: Throwable) {
            return Promise.rejected(Throwable("SDK is not initialized, call initialize() first"))
        }

        val eventType = when (eventType) {
            UserComProductEventType.ADDTOCART -> ProductEventType.ADD_TO_CART
            UserComProductEventType.PURCHASE -> ProductEventType.PURCHASE
            UserComProductEventType.LIKING -> ProductEventType.LIKING
            UserComProductEventType.ADDTOOBSERVATION -> ProductEventType.ADD_TO_OBSERVATION
            UserComProductEventType.ORDER -> ProductEventType.ORDER
            UserComProductEventType.RESERVATION -> ProductEventType.RESERVATION
            UserComProductEventType.RETURN -> ProductEventType.RETURN
            UserComProductEventType.VIEW -> ProductEventType.VIEW
            UserComProductEventType.CLICK -> ProductEventType.CLICK
            UserComProductEventType.DETAIL -> ProductEventType.DETAIL
            UserComProductEventType.ADD -> ProductEventType.ADD
            UserComProductEventType.REMOVE -> ProductEventType.REMOVE
            UserComProductEventType.CHECKOUT -> ProductEventType.CHECKOUT
            UserComProductEventType.CHECKOUTOPTION -> ProductEventType.CHECKOUT_OPTION
            UserComProductEventType.REFUND -> ProductEventType.REFUND
            UserComProductEventType.PROMOCLICK -> ProductEventType.PROMO_CLICK
        }

        instance.sendProductEvent(productId, eventType, params?.toHashMap())
        return Promise.resolved()
    }

    override fun sendCustomEvent(
        eventName: String,
        data: AnyMap
    ): Promise<Unit> {
        val instance = try {
            UserCom.getInstance()
        } catch (_: Throwable) {
            return Promise.rejected(Throwable("SDK is not initialized, call initialize() first"))
        }

        instance.sendEvent(eventName, data.toHashMap())
        return Promise.resolved()
    }

    override fun sendScreenEvent(screenName: String): Promise<Unit> {
        val instance = try {
            UserCom.getInstance()
        } catch (_: Throwable) {
            return Promise.rejected(Throwable("SDK is not initialized, call initialize() first"))
        }

        instance.trackScreen(screenName)
        return Promise.resolved()
    }
}
